### PR TITLE
Feat/improve scalability

### LINKED ISSUE #17: Duplicate key value violates unique constraint "sessions_name__key"
### Description

Randomly I get an error message like:

```
error: duplicate key value violates unique constraint "sessions_name__key"
    at c:\myapp\node_modules\pg-pool\index.js:45:11
    at process.processTicksAndRejections (C:\myapp\lib\internal\process\task_queues.js:95:5)
    at async createNewSession (c:\myapp\node_modules\@acpr\rate-limit-postgresql\dist\index.mjs:53:18)
    at async PostgresStoreAggregatedIP.increment (c:\myapp\node_modules\@acpr\rate-limit-postgresql\dist\index.mjs:304:22)
    at async file:///C:/myapp/node_modules/express-rate-limit/dist/index.mjs:602:40
    at async file:///C:/myapp/node_modules/express-rate-limit/dist/index.mjs:583:5 {length: 263, name: 'error', severity: 'ERROR', code: '23505', detail: 'Key (name_)=(aggregated_store) already exists.', …}
```

My Rate Limiter is setup as:

```
import { rateLimit } from 'express-rate-limit'
import { PostgresStore } from '@acpr/rate-limit-postgresql'

const apiRateLimiter = rateLimit({
  store: new PostgresStore(
    {
      user: process.env.DB_USER,
      password: process.env.DB_PASSWORD,
      host: process.env.DB_HOST,
      database: process.env.DB_DATABASE,
      port: parseInt(process.env.DB_PORT)
    },
    'aggregated_store'
  ),
  windowMs: 1 * 60 * 1000, // 1 minute
  limit: 600, // limit each IP/email to 600 requests per windowMs (1 minute) - 10 requests per second
  delayMs: 0, // disable delaying - user has full speed until the max limit is reached
  standardHeaders: 'draft-7', // draft-6: `RateLimit-*` headers; draft-7: combined `RateLimit` header
  legacyHeaders: false, // Disable the `X-RateLimit-*` headers.
  keyGenerator: (req, res) => {
    if (req?.currentUser?.email) {
      return req.currentUser.email
    } else {
      return req.ip
    }
  }
})
```

### Library version

@acpr/rate-limit-postgresql: ^1.3.2
express-rate-limit: ^7.1.5

### Node version

20.11.0

### Typescript version (if you are using it)

_No response_

### Module system

ESM

### PR DESCRIPTION
**Pull request by adrianprelipcean**
# Description 

Moves the session handling logic on the database side to minimize concurrency errors. 

Tests are performed on artillery using the following setup, where `users-*` and `users-*-shared` use the same store, and `users-*-unique` use a separate store (to check for false positives and edge cases around reusability.) Also `agg` are endpoints using an aggregated (default) store, and `ind` are endpoints using an individual store. 

```yaml
config:
  target: "http://localhost:3005"
  phases:
    - duration: 30s
      arrivalRate: 1
      rampTo: 100
      name: ramp up
    - duration: 20s
      arrivalRate: 100
      name: sustain

scenarios:
  - flow:
      - loop:
        - get:
            url: "/users-agg"
        - get:
            url: "/users-agg-shared"
        - get:
            url: "/users-agg-unique"
        - get:
            url: "/users-ind"
        - get:
            url: "/users-ind-shared"
        - get:
            url: "/users-ind-unique"
        count: 5
``` 

### Results on old version 

```
errors.ETIMEDOUT: .............................................................. 3020
http.codes.200: ................................................................ 3146
http.codes.429: ................................................................ 13682
http.codes.500: ................................................................ 7640
http.downloaded_bytes: ......................................................... 6608600
http.request_rate: ............................................................. 357/sec
http.requests: ................................................................. 27488
http.response_time:
  min: ......................................................................... 0
  max: ......................................................................... 9998
  mean: ........................................................................ 576.8
  median: ...................................................................... 26.8
  p95: ......................................................................... 4965.3
  p99: ......................................................................... 9416.8
http.responses: ................................................................ 24468
vusers.completed: .............................................................. 495
vusers.created: ................................................................ 3515
vusers.created_by_name.0: ...................................................... 3515
vusers.failed: ................................................................. 3020
vusers.session_length:
  min: ......................................................................... 66.1
  max: ......................................................................... 39913.2
  mean: ........................................................................ 11332.7
  median: ...................................................................... 2671
  p95: ......................................................................... 31266.3
  p99: ......................................................................... 33870.4
```

### Results on new version after changes 
| Note: timeouts default to 10s in artillery 
```
errors.ETIMEDOUT: .............................................................. 999
http.codes.200: ................................................................ 4717
http.codes.429: ................................................................ 77717
http.downloaded_bytes: ......................................................... 5866738
http.request_rate: ............................................................. 975/sec
http.requests: ................................................................. 83433
http.response_time:
  min: ......................................................................... 0
  max: ......................................................................... 10058
  mean: ........................................................................ 885.2
  median: ...................................................................... 77.5
  p95: ......................................................................... 7407.5
  p99: ......................................................................... 9999.2
http.responses: ................................................................ 82434
vusers.completed: .............................................................. 2516
vusers.created: ................................................................ 3515
vusers.created_by_name.0: ...................................................... 3515
vusers.failed: ................................................................. 999
vusers.session_length:
  min: ......................................................................... 69.2
  max: ......................................................................... 42172.4
  mean: ........................................................................ 26288.9
  median: ...................................................................... 35964.9
  p95: ......................................................................... 42205.5
  p99: ......................................................................... 42205.5
```

### Reference for in memory 
| Note - in memory has higher throughput (2x compared to new version), which is expected in the current setup. 
```
http.codes.200: ................................................................ 4332
http.codes.429: ................................................................ 101118
http.downloaded_bytes: ......................................................... 7559226
http.request_rate: ............................................................. 2078/sec
http.requests: ................................................................. 105450
http.response_time:
  min: ......................................................................... 0
  max: ......................................................................... 426
  mean: ........................................................................ 26.6
  median: ...................................................................... 22.9
  p95: ......................................................................... 82.3
  p99: ......................................................................... 117.9
http.responses: ................................................................ 105450
vusers.completed: .............................................................. 3515
vusers.created: ................................................................ 3515
vusers.created_by_name.0: ...................................................... 3515
vusers.failed: ................................................................. 0
vusers.session_length:
  min: ......................................................................... 23.7
  max: ......................................................................... 2977.3
  mean: ........................................................................ 823.5
  median: ...................................................................... 772.9
  p95: ......................................................................... 2515.5
  p99: ......................................................................... 2780
```

Fixes #17 , #15

### CODE REVIEW

**Comment by nfriedly** [comment]
That's quite a significant improvement in throughput, and no 500 errors!

I'm reading through the changes, but I may not finish reviewing it today.

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@acpr/rate-limit-postgresql",
-  "version": "1.3.2",
```
Comment (nfriedly) [inline package.json:3]: I usually do this separately by running `npm version minor` (or `major` or `patch`) on the main branch after the PR is merged in. 

The npm command creates both a commit and a git tag, and then you can push both with `git push --follow-tags`. The tag makes it much easier to find the code for a specific version later on.

**Review by nfriedly — APPROVED** [review]
This looks really good to me. I learned a bit just reading your SQL, so I'm obviously not going to be the most critical reviewer, but I think the general ideal of letting the DB handle the sessions is correct, and the results indicate to me that this is both a solid implementation and a significant improvement.

Thanks for doing it!
